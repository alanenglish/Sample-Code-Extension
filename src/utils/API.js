const CATEGORIES_URL = 'https://api.reverb.com/api/categories/flat';
const LISTINGS_URL = 'https://api.reverb.com/api/listings';
const CATEGORY_LISTINGS_URL = 'https://api.reverb.com/api/listings?category_uuid=';

const HEADERS = {
  Accept: 'application/json',
  'Accept-Version': '3.0',
  'Content-Type': 'application/json'
};

/**
 *  getJSON method always returns an Object
 */

async function getJSON(url, currency = 'USD', region = 'XX') {
  const response = await fetch(url, { headers: { ...HEADERS, 'X-Display-Currency': currency, 'X-Shipping-Region': region } });
  return await response.json();
}

export async function fetchCategories() {
  return getJSON(CATEGORIES_URL);
}

export async function fetchListings(page = 1, perPage = 10, currency = 'USD') {
  return getJSON(`${LISTINGS_URL}/?page=${page}&per_page=${perPage}`, currency);
}

export async function fetchCategoryListings(id, currency = 'USD', region = 'XX') {
  return getJSON(`${CATEGORY_LISTINGS_URL}${id}`, currency, region);
}

export async function fetchListing(id) {
  return getJSON(`${LISTINGS_URL}/${id}`);
}

// export async function fetchListings() {
//   return getJSON(LISTINGS_URL);
// }

// ERROR HANDLING OPTIONS
// if (response.status !== 200) {
//   return { error: true };
// }

// if (response.status !== 200) {
//   let message = { message: 'default error' };
//   try {
//     message = await response.json();
//   } catch (error) {
//     // does nothing
//   }
//   return message;
// }
