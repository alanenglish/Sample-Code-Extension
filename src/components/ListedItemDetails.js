import React from 'react';
import PropTypes from 'prop-types';
import numeral from 'numeral';
import { fetchListing } from '../utils/API';

class ListedItemDetail extends React.Component {
  static propTypes = {
    match: PropTypes.object,
    history: PropTypes.object.isRequired
  }

  static defaultProps = {
    match: {}
  }

  state = {
    listing: null
  }

  async componentDidMount() {
    const response = await fetchListing(this.props.match.params.id);

    this.setState({
      listing: response
    });
  }

  render() {
    const { listing } = this.state;

    return (
      <div className="content-container">
        <div>
          <button className="back-btn" onClick={() => this.props.history.goBack()}>Back</button>
        </div>
        { this.state.listing &&
          <div className="category-show">
            <h1 className="category-show__header">{listing.title}</h1>
            <div className="row">
              <div className="col-lg-5 col-md-5 col-sm-12 left-col">
                <img
                  className="category-show__img"
                  alt={listing.title}
                  src={listing.photos[0]._links.full.href}
                />
              </div>
              <div className="col-lg-7 col-md-7 col-sm-12 right-col">
                <div className="category-show__attribute">
                  Make: <span className="make">{listing.make}</span>
                </div>
                <div className="category-show__attribute">
                  Model: <span className="model">{listing.model}</span>
                </div>
                { this.state.listing.year &&
                  <div className="category-show__attribute">
                    Year: <span className="year">{listing.year}</span>
                  </div>
                }
                { this.state.listing.finish &&
                  <div className="category-show__attribute">
                    Finish: <span className="finish">{listing.finish}</span>
                  </div>
                }
                <div className="category-show__attribute">
                  Description: <span className="description">{listing.description}</span>
                </div>
                <div className="category-show__attribute">
                  Price: <span className="price">{numeral(listing.price.amount_cents / 100).format('$0,0.00')}</span>
                </div>
                <div className="category-show__attribute">
                  Condition: <span className="condition">{listing.condition.display_name}</span>
                </div>
                <div className="category-show__attribute">
                  Shop Name: <span className="shop-name">{listing.shop_name}</span>
                </div>
                { this.state.listing.shop.preferred_seller &&
                  <div className="category-show__attribute">
                    Preferred Seller: <span className="preferred-seller">Yes</span>
                  </div>
                }
                <div className="category-show__attribute">
                  Return Policy: <span className="return-policy">{listing.return_policy.description}</span>
                </div>
              </div>
            </div>
          </div>
        }
      </div>
    );
  }
}

export default ListedItemDetail;

// console.log(props.match.params.id);
// props.location.state
