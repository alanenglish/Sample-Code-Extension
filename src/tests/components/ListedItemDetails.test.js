import React from 'react';
import { mount, shallow } from 'enzyme';
import * as API from '../../utils/API';
import ListedItemDetails from '../../components/ListedItemDetails';
/* eslint-env jest */

describe('<ListedItemDetails />', () => {
  let response;
  let history;

  beforeEach(() => {
    response = Promise.resolve({
      listing: {
        id: '123abc456def',
        title: 'Fender Guitar Brand New',
        make: 'Fender',
        model: 'FSR Stratocaster XII 12 string',
        year: '2017',
        finish: 'Sunburst',
        description: 'Immaculate condition guitar, you will not be disappointed!',
        price: 67500,
        condition: 'Mint',
        shop_name: 'The Official Awesome Reverb Shop',
        preferred_seller: true,
        return_policy: 'This item is sold As-Described and cannot be returned unless it arrives in a condition different from how it was described or photographed.',
        photos: [{ _links: { small_crop: { href: 'http://www.alan.com' } } }]
      }
    });

    history = { goBack: jest.fn() };

    jest.spyOn(API, 'fetchListing').mockImplementation(() => (response));
  });

  it('displays single listing from the API on mount', () => {
    // const page = mount(<ListedItemDetails history={history} />);
    // await response;
    // page.update();
    //
    // expect(page.find('h1').at(0).text()).toEqual('Fender Guitar Brand New');
    const wrapper = shallow(<ListedItemDetails history={history} />);
    const listing = {
      id: '123abc456def',
      title: 'Fender Guitar Brand New',
      make: 'Fender',
      model: 'FSR Stratocaster XII 12 string',
      year: '2017',
      finish: 'Sunburst',
      description: 'Immaculate condition guitar, you will not be disappointed!',
      price: { amount_cents: 67500 },
      condition: { display_name: 'Mint' },
      shop_name: 'The Official Awesome Reverb Shop',
      shop: { preferred_seller: true },
      return_policy: { description: 'This item is sold As-Described and cannot be returned unless it arrives in a condition different from how it was described or photographed.' },
      photos: [{ _links: { full: { href: 'http://www.alan.com' } } }]
    };

    wrapper.setState({ listing: listing });
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
});
