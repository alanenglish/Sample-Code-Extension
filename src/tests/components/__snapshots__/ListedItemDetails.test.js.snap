// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListedItemDetails /> displays single listing from the API on mount 1`] = `
<div
  className="content-container"
>
  <div>
    <button
      className="back-btn"
      onClick={[Function]}
    >
      Back
    </button>
  </div>
  <div
    className="category-show"
  >
    <h1
      className="category-show__header"
    >
      Fender Guitar Brand New
    </h1>
    <div
      className="row"
    >
      <div
        className="col-lg-5 col-md-5 col-sm-12 left-col"
      >
        <img
          alt="Fender Guitar Brand New"
          className="category-show__img"
          src="http://www.alan.com"
        />
      </div>
      <div
        className="col-lg-7 col-md-7 col-sm-12 right-col"
      >
        <div
          className="category-show__attribute"
        >
          Make: 
          <span
            className="make"
          >
            Fender
          </span>
        </div>
        <div
          className="category-show__attribute"
        >
          Model: 
          <span
            className="model"
          >
            FSR Stratocaster XII 12 string
          </span>
        </div>
        <div
          className="category-show__attribute"
        >
          Year: 
          <span
            className="year"
          >
            2017
          </span>
        </div>
        <div
          className="category-show__attribute"
        >
          Finish: 
          <span
            className="finish"
          >
            Sunburst
          </span>
        </div>
        <div
          className="category-show__attribute"
        >
          Description: 
          <span
            className="description"
          >
            Immaculate condition guitar, you will not be disappointed!
          </span>
        </div>
        <div
          className="category-show__attribute"
        >
          Price: 
          <span
            className="price"
          >
            $675.00
          </span>
        </div>
        <div
          className="category-show__attribute"
        >
          Condition: 
          <span
            className="condition"
          >
            Mint
          </span>
        </div>
        <div
          className="category-show__attribute"
        >
          Shop Name: 
          <span
            className="shop-name"
          >
            The Official Awesome Reverb Shop
          </span>
        </div>
        <div
          className="category-show__attribute"
        >
          Preferred Seller: 
          <span
            className="preferred-seller"
          >
            Yes
          </span>
        </div>
        <div
          className="category-show__attribute"
        >
          Return Policy: 
          <span
            className="return-policy"
          >
            This item is sold As-Described and cannot be returned unless it arrives in a condition different from how it was described or photographed.
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
